doc/design/refactor/multi_device.md

```diff
@@ -0,0 +1,37 @@
+# Design Doc: Execute the Program with Multi Thread
+
+## Abstract
+
+This Design Doc propose an approach to make the user-defined Op graph
+running with multi threads, we will use an auto transpiler to convert the user-defined
+Op graph to a multi-thread Op graph, and run `ParallelDo` Op to run the
+multi-thread graph.
+
+## Graph Converter
+<img src="src/multi-threads/single-thread@3x.png" width="300">
+
+After converted:
+
+<img src="src/multi-threads/multi-threads@3x.png" width="1000">
+
+## Implement
+
+- `Multi-Thread Transpiler`  will convert the graph to a multi-threads graph
+    which would be executed with multi-threads.
+- `BlockingCounter` will `Init/Decrement` a condition variable, and Blocking `Wait`
+    for the condition variable become `0`:
+    ```cpp
+    BlockingCounter bc(thread_count);
+    for (int i = 0; i < thread_count; ++i) {
+        thread_pool->Schedule([&bc] {bc.DecrementCount(); })
+    }
+    bc.Wait();
+    ```
+- `ParallelDo` Operator
+  - Initialize a thread pool which is a Singleton.
+  - Use a list of block id as the input, and create multi Executor to run 
+    these Blocks with multi-threads.
+  - Initialize a `BlockingCounter` instance and wait until all threads are done.
+- `Split` Operator will split the Input Tensor into N slices.
```

Comment:
Can split and merge reuse current operators like split_op and mean_op to merge?

Reply:
Maybe we need to do something enhancement such as how to split a LodTensor, and Merge maybe  including mean and other mathematical methods such as *sum/max* ..., but you're right,  `mean_op/split_op` would be reused.